src/NuGet.Clients/PackageManagement.UI/Common/PackageCollection.cs

```diff
@@ -13,40 +13,42 @@ namespace NuGet.PackageManagement.UI
     /// <summary>
     /// Wrapper class consolidating common queries against a collection of packages
     /// </summary>
-    internal class PackageCollection : IEnumerable<PackageIdentity>
+    internal class PackageCollection : IEnumerable<PackageCollectionItem>
     {
-        private readonly PackageIdentity[] _packages;
+        private readonly PackageCollectionItem[] _packages;
         private readonly ISet<string> _uniqueIds = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
 
-        public PackageCollection(PackageIdentity[] packages)
+        public PackageCollection(PackageCollectionItem[] packages)
         {
             _packages = packages;
             _uniqueIds.UnionWith(_packages.Select(p => p.Id));
         }
 
-        public IEnumerator<PackageIdentity> GetEnumerator()
+        public IEnumerator<PackageCollectionItem> GetEnumerator()
         {
-            return ((IEnumerable<PackageIdentity>)_packages).GetEnumerator();
+            return ((IEnumerable<PackageCollectionItem>)_packages).GetEnumerator();
         }
 
         IEnumerator IEnumerable.GetEnumerator()
         {
-            return ((IEnumerable<PackageIdentity>)_packages).GetEnumerator();
+            return ((IEnumerable<PackageCollectionItem>)_packages).GetEnumerator();
         }
 
         public bool ContainsId(string packageId) => _uniqueIds.Contains(packageId);
 
         public static async Task<PackageCollection> FromProjectsAsync(IEnumerable<NuGetProject> projects, CancellationToken cancellationToken)
         {
+            // Read package references from all projects.
             var tasks = projects
                 .Select(project => project.GetInstalledPackagesAsync(cancellationToken));
             var packageReferences = await Task.WhenAll(tasks);
+
+            // Group all package references for an id/version into a single item.
             var packages = packageReferences
-                .SelectMany(p => p)
-                .Where(p => p != null)
-                .Select(p => p.PackageIdentity)
-                .Distinct(PackageIdentity.Comparer)
-                .ToArray();
+                    .SelectMany(e => e)
+                    .GroupBy(e => e.PackageIdentity)
```

Comment:
can join `group by` and `select` into a single `group by`

`.GroupBy(e => e.PackageIdentity, (key, group) =>new PackageCollectionItem(key.Id, key.Version, group) )`